src/OpenTelemetry.Contrib.Instrumentation.Elasticsearch/Implementation/ElasticsearchRequestPipelineDiagnosticListener.cs

```diff
@@ -0,0 +1,207 @@
+using System.IO;
+using System.Linq;
+using System.Net.Http;
+using System.Net.Sockets;
+using System.Text;
+using System.Text.Json;
+using System.Text.RegularExpressions;
+using OpenTelemetry.Instrumentation;
+using OpenTelemetry.Trace;
+
+namespace OpenTelemetry.Contrib.Instrumentation.ElasticsearchClient.Implementation
+{
+    internal class ElasticsearchRequestPipelineDiagnosticListener : ListenerHandler
+    {
+        private static readonly Regex ParseRequest = new Regex(@"\n# Request:\r?\n(.*)\n# Response", RegexOptions.Compiled | RegexOptions.Singleline);
+
+        private readonly ActivitySourceAdapter activitySource;
+        private readonly ElasticsearchClientInstrumentationOptions options;
+        private readonly PropertyFetcher<Uri> uriFetcher = new PropertyFetcher<Uri>("Uri");
+        private readonly PropertyFetcher<object> methodFetcher = new PropertyFetcher<object>("Method");
+        private readonly PropertyFetcher<bool> successFetcher = new PropertyFetcher<bool>("SuccessOrKnownError");
+        private readonly PropertyFetcher<string> debugInformationFetcher = new PropertyFetcher<string>("DebugInformation");
+        private readonly PropertyFetcher<int?> httpStatusFetcher = new PropertyFetcher<int?>("HttpStatusCode");
+        private readonly PropertyFetcher<Exception> orginalExceptionFetcher = new PropertyFetcher<Exception>("OriginalException");
+
+        public ElasticsearchRequestPipelineDiagnosticListener(ActivitySourceAdapter activitySource, ElasticsearchClientInstrumentationOptions options)
+            : base("Elasticsearch.Net.RequestPipeline")
+        {
+            if (activitySource == null)
+            {
+                throw new ArgumentNullException(nameof(activitySource));
+            }
+
+            this.activitySource = activitySource;
+            this.options = options;
+        }
+
+        public override void OnStartActivity(Activity activity, object payload)
+        {
+            var uri = this.uriFetcher.Fetch(payload);
+
+            if (uri == null)
+            {
+                ElasticsearchInstrumentationEventSource.Log.NullPayload(nameof(ElasticsearchRequestPipelineDiagnosticListener), nameof(this.OnStartActivity));
+                return;
+            }
+
+            var method = this.methodFetcher.Fetch(payload)?.ToString();
+
+            if (activity.OperationName == "Ping")
+            {
+                activity.DisplayName = $"Elasticsearch: Ping";
+            }
+            else
+            {
+                activity.DisplayName = $"Elasticsearch: {method} {uri.AbsolutePath}";
+            }
+
+            this.activitySource.Start(activity, ActivityKind.Client);
+
+            if (this.options.SuppressDownstreamInstrumentation)
+            {
+                SuppressInstrumentationScope.Enter();
+            }
+
+            if (activity.IsAllDataRequested)
+            {
+                activity.AddTag(Constants.AttributeDbSystem, "elasticsearch");
```

Comment:
use `SetTag` all the time :)
`AddTag` is part of activity, but `SetTag` is the one which behaves like OpenTelemetry spec :)

Reply:
Updated